### ISSUE TITLE
@doc and @inline bad interaction

### ISSUE THREAD

**Issue by Luthaf** [JuliaLang/julia#11443]
The following code 
```
julia> @doc "The `foo` function" ->
       @inline function foo()
           42
       end
(anonymous function)
```

Give me an anonymous function, which then lead to further errors:

```

julia> foo(4)
ERROR: UndefVarError: foo not defined
```

Some more information
```
julia> versioninfo()
Julia Version 0.4.0-dev+4984
Commit 32f5a51* (2015-05-25 06:56 UTC)
Platform Info:
  System: Darwin (x86_64-apple-darwin13.4.0)
  CPU: Intel(R) Core(TM) i7-3520M CPU @ 2.90GHz
  WORD_SIZE: 64
  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Sandybridge)
  LAPACK: libopenblas
  LIBM: libopenlibm
  LLVM: libLLVM-3.3
```

**Comment by one-more-minute** [comment]
I guess the combination of `@doc` and `@inline` ends up producing a new scope somehow. What does the macro expansion looks like?

**Comment by Luthaf** [comment]
Like this : 

```
julia> :(@doc "The `foo` function" ->
              @inline function foo()
                  42
              end)
:(@doc "The `foo` function"->begin  # none, line 1:
                @inline function foo() # none, line 3:
                        42
                    end
            end)

julia> macroexpand(ans)
quote  # docs.jl, line 182:
    begin  # docs.jl, line 18:
        if Base.Docs.!(Base.Docs.isdefined(:META)) # line 19:
            const META = Base.Docs.ObjectIdDict() # line 20:
            Base.Docs.doc!(META,begin  # markdown/Markdown.jl, line 42:
                    let #7#md = (Base.Markdown.MD)(Any[(Base.Markdown.Paragraph)(Any["Documentation metadata for ",Base.Markdown.Code("","Main"),"."])],Dict{Any,Any}(:config=>Base.Markdown.Config([hashheader,fencedcode,blockquote],[blocktex,blockinterp,list,indentcode,github_table,horizontalrule,setextheader,paragraph],Dict('['=>[link],'\\'=>[linebreak,escapes],'*'=>[asterisk_bold,asterisk_italic],'$'=>[tex,interp],'`'=>[inline_code],'!'=>[image],'-'=>[en_dash])))) # line 43:
                        #7#md.meta[:path] = nothing # line 44:
                        #7#md.meta[:module] = Base.Markdown.current_module() # line 45:
                        #7#md
                    end
                end) # line 21:
            Base.Docs.push!(Base.Docs.modules,Base.Docs.current_module()) # line 22:
            Base.Docs.nothing
        end
    end # line 183:
    #6#f = @inline # line 184:
    Base.Docs.doc!(#6#f,begin  # markdown/Markdown.jl, line 42:
            let #9#md = (Base.Markdown.MD)(Any[(Base.Markdown.Paragraph)(Any["The ",Base.Markdown.Code("","foo")," function"])],Dict{Any,Any}(:config=>Base.Markdown.Config([hashheader,fencedcode,blockquote],[blocktex,blockinterp,list,indentcode,github_table,horizontalrule,setextheader,paragraph],Dict('['=>[link],'\\'=>[linebreak,escapes],'*'=>[asterisk_bold,asterisk_italic],'$'=>[tex,interp],'`'=>[inline_code],'!'=>[image],'-'=>[en_dash])))) # line 43:
                #9#md.meta[:path] = nothing # line 44:
                #9#md.meta[:module] = Base.Markdown.current_module() # line 45:
                #9#md
            end
        end) # line 185:
    #6#f
end
```

May the `#6#f = @inline # line 184:` line be the culprit here ?

**Comment by mauro3** [comment]
I have the same problem with the `@traitdef` macro in Traits.jl:
```
julia> using Traits
  This warning is ok:
Warning: Method definition func_for_method(Method, Any, Any) in module Inference at inference.jl:593 overwritten in module Traits at /home/mauro/.julia/v0.4/Traits/src/base_fixes.jl:6.
  endof ok-warning.

julia> macroexpand(:(@traitdef Tr{X} begin
       sin(X)
       end))
:(immutable Tr{X} <: Traits.Trait{Tuple{}} # /home/mauro/.julia/v0.4/Traits/src/traitdef.jl, line 270:
        methods::Traits.FDict # line 271:
        constraints::Vector{Bool} # line 272:
        assoctyps::Vector{Any} # line 273:
        function Tr() # /home/mauro/.julia/v0.4/Traits/src/traitdef.jl, line 274:
            begin 
                assoctyps = Any[]
            end # line 275:
            new(Traits.FDict(sin => __sin(::X) = Any()),Bool[],assoctyps)
        end
    end)

julia> macroexpand(:(@doc "Help!" @traitdef Tr{X} begin
       sin(X)
       end))
quote  # docs.jl, line 182:
    begin  # docs.jl, line 18:
        if Base.Docs.!(Base.Docs.isdefined(:META)) # line 19:
            const META = Base.Docs.ObjectIdDict() # line 20:
            Base.Docs.doc!(META,begin  # markdown/Markdown.jl, line 42:
                    let #197#md = (Base.Markdown.MD)(Any[(Base.Markdown.Paragraph)(Any["Documentation metadata for ",Base.Markdown.Code("","Main"),"."])],Dict{Any,Any}(:config=>Base.Markdown.Config([hashheader,fencedcode,blockquote],[blocktex,blockinterp,list,indentcode,github_table,horizontalrule,setextheader,paragraph],Dict('['=>[link],'\\'=>[linebreak,escapes],'*'=>[asterisk_bold,asterisk_italic],'$'=>[tex,interp],'`'=>[inline_code],'!'=>[image],'-'=>[en_dash])))) # line 43:
                        #197#md.meta[:path] = nothing # line 44:
                        #197#md.meta[:module] = Base.Markdown.current_module() # line 45:
                        #197#md
                    end
                end) # line 21:
            Base.Docs.push!(Base.Docs.modules,Base.Docs.current_module()) # line 22:
            Base.Docs.nothing
        end
    end # line 183:
    #196#f = @traitdef # line 184:
    Base.Docs.doc!(#196#f,begin  # markdown/Markdown.jl, line 42:
            let #199#md = (Base.Markdown.MD)(Any[(Base.Markdown.Paragraph)(Any["Help!"])],Dict{Any,Any}(:config=>Base.Markdown.Config([hashheader,fencedcode,blockquote],[blocktex,blockinterp,list,indentcode,github_table,horizontalrule,setextheader,paragraph],Dict('['=>[link],'\\'=>[linebreak,escapes],'*'=>[asterisk_bold,asterisk_italic],'$'=>[tex,interp],'`'=>[inline_code],'!'=>[image],'-'=>[en_dash])))) # line 43:
                #199#md.meta[:path] = nothing # line 44:
                #199#md.meta[:module] = Base.Markdown.current_module() # line 45:
                #199#md
            end
        end) # line 185:
    #196#f
end

julia> @doc "Help!" @traitdef Tr{X} begin
       sin(X)
       end
(anonymous function)
```

**Comment by yuyichao** [comment]
Related https://github.com/JuliaLang/julia/issues/10887

**Comment by jakebolewski** [comment]
This seems to still be an issue.

**Comment by timholy** [comment]
You also can't use `@doc` before a `@generated` function. However, you can use it before a (deprecated) `stagedfunction`.

**Comment by yuyichao** [comment]
Close in favor of #11993 since that one is more general.

Both should be fixed by #12000
